File: 01_Language/01_Functions/java/src/main/java/com/onevgo/functions/ArrayRand.java
package com.onevgo.functions;

import java.util.*;
import java.util.concurrent.ThreadLocalRandom;
import java.util.stream.IntStream;

public class ArrayRand {
    public static <E> List<Integer> arrayRand(List<E> list, int num) {
        if (num < 1 || num > list.size()) {
            throw new IllegalArgumentException("num");
        }

        List<Integer> result = new ArrayList<>(num);
        // ranges = [ 0, list.size() )
        int[] ranges = IntStream.range(0, list.size()).toArray();
        int len = ranges.length;
        for (int i = 0; i < num; i++) {
            int p = ThreadLocalRandom.current().nextInt(len - i);
            result.add(ranges[p]);
            ranges[p] = ranges[len - 1 - i];
        }
        return result;
    }

    public static void main(String[] args) {
        List<String> list = Arrays.asList("Neo", "Morpheus", "Trinity", "Cypher", "Tank");
        System.out.println(arrayRand(list, 2));
        System.out.println(arrayRand(list, 2));
        System.out.println(arrayRand(list, 2));
        System.out.println(arrayRand(list, 2));
        System.out.println(arrayRand(list, 2));
    }
}

File: 01_Language/01_Functions/js/array_rand.js
function array_rand(arr, num = 1) {
    let res = []
    let keys = Object.keys(arr)
    let endIndex = keys.length - 1
    for (let i = 0; i < num; i++) {
        let targetIndex = Math.floor(Math.random() * (endIndex + 1))
        res.push(keys[targetIndex])
        keys[targetIndex] = keys[endIndex]
        endIndex--
    }
    return num === 1 ? res[0] : res
}

let input = ["Neo", "Morpheus", "Trinity", "Cypher", "Tank"]
console.log(array_rand(input, 2));

console.log(array_rand(input));

input = {"k1": "Neo", "k2": "Morpheus", "k3": "Trinity", "k4": "Cypher", "k5": "Tank"}
console.log(array_rand(input, 2));
